### PR TITLE
Ampers: Abinnet ride share waves 1- 3

### PR DESCRIPTION
**Pull request by Abiaina**
…ted the driver, trips etc to make start/end times an instance of time class.

# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to figure out how I would track new trips and how to update status of drivers as I assigned them to a new trip.  I ended up making driver status an attr_accessor so it could be changed dispatcher creates a new trip and assigns a driver.
Describe any examples of composition that you encountered in this project, if any | There was many. All drivers and passengers had trips. And dispatcher had access to drivers, passengers and trip instance methods/variables. 
Describe a _nominal test_ that you wrote for this assignment. | I tested that the first and 3rd selected driver for the wave 3 requirement were the ones outlined by dee in slack. I expected a driver without any past trips to be selected first and then a driver with a trip end_time the furthest in the pass to be selected after.
Describe an _edge case test_ that you wrote for this assignment. | I tested what would happen if there are no more available drivers. I confirmed that the method getting the available drivers still returned an array and wrote a test that requested a trip for each available driver. When the available drivers method was called it returned an array of the value nil (an empty array). This tested the limit of how many request trips can be called and drivers can be assigned to the newly requested trip.
Describe a concept that you gained more clarity on as you worked on this assignment. | I definitely began to understand the difference between attr_reader and attr_accessor. I am beginning to understand how composition uses these to help control  the change in value of instance variables as they are called in another class.

### CODE REVIEW

File: lib/passenger.rb [hunk 1]
```diff
@@ -20,5 +22,25 @@ def get_drivers
     def add_trip(trip)
       @trips << trip
     end
+
+    def trips_total
+      total = 0
+      @trips.each do |trip|
+        if trip.duration != :nil
```
Comment (CheezItMan) [inline lib/passenger.rb:29]: symbol `nil`?

File: lib/driver.rb [hunk 2]
```diff
@@ -43,5 +45,36 @@ def add_trip(trip)
...
+      income = 0
+      @trips.each do |trip|
+        if trip.duration != nil
+          income += trip.cost
+        end
+      end
+      if income == 0
+        net_income = income
+      else
+        net_income = (income - 1.65) * 0.8
```
Comment (CheezItMan) [inline lib/driver.rb]: You may also want to include a check that income > 1.65, most likely, but worth a check.

File: lib/trip_dispatcher.rb [hunk 3]
```diff
@@ -87,7 +95,64 @@ def load_trips
...
+      free_drivers = []
+      @drivers.each do |driver|
+        if driver.status == :AVAILABLE
+          free_drivers << driver
+        end
+      end
+      return free_drivers
+    end
+
+    def last_trip (driver)
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb]: This method would make more sense as a member of the `Driver` class.

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -87,7 +95,64 @@ def load_trips
...
+
+    def select_driver
+      free_drivers = available_drivers
+      driver_last_trip = {}
+      free_drivers.each do |driver|
+        if driver.trips.count == 0
+          return driver
+        elsif
+          last_trip = last_trip(driver)
+          driver_last_trip.store(driver,last_trip)
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:115]: what is this doing here?

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -87,7 +95,64 @@ def load_trips
...
+      if driver != nil
+        driver.status = :UNAVAILABLE
+        new_trip[:driver] = driver
+        new_trip[:passenger] = find_passenger(passenger_id)
+        new_trip[:start_time] = start_time
+        new_trip[:end_time] = nil
+        new_trip[:cost] = nil
+        new_trip[:rating] = nil
+        newly_added_trip = RideShare::Trip.new(new_trip)
+        @trips << newly_added_trip
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:139]: You are not adding the trip to the `Driver` and `Passenger` as well.

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -89,4 +91,84 @@
...
+      @test_trip.start_time.must_be_kind_of Time
+    end
+
+    it "request_trip assigned driver status is UNAVAILABLE" do
+      @test_trip.driver.status.must_equal :UNAVAILABLE
+    end
+
+    it "available drivers count goes down by 1 after request_ride is called" do
+      test_dispatcher = RideShare::TripDispatcher.new
+      before = test_dispatcher.available_drivers.count
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb]: `before` is a word used in minitest, use instead `before_driver_count`

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -89,4 +91,84 @@
...
+      test_trip = test_dispatcher.request_trip(5)
+      after = test_dispatcher.available_drivers.count
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb]: `after` is also an issue

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -89,4 +91,84 @@
       passenger.trips.must_include trip
     end
   end
+
+  describe "request_trip" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:95]: Missing test:
You also need to test what happens when you try to request trips when there are no available drivers.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | More commits would be better, and don't focus on waves for commit messages, describe the functionality you added.
Answer comprehension questions |  Check, I'm glad you gained more clarity in composition and `attr_reader` and `attr_accessor`.
**Wave 1** |
`Trip` has a helper method to calculate duration |  Check, well done
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue |  Check
Complete tests |  Good tests and good work taking the edge cases for duration.
**Wave 2** |
`TripDispatcher` has a new method to create trips | Check, see my notes, you're not adding the newly created trip to the corresponding driver and passenger.  
creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Complete tests |  **Some issues here** See my in-code notes
**Wave 3** |
Complex logic was correctly implemented and tested | **INCOMPLETE**
Appropriate use of Ruby's `Date` |  Good use of `Time`
Appropriate helper methods were made to help with complex logic | NA
Complete tests | NA
**Overall** | You hit **most** of the learning goals.  See my inline notes.  You missed a few things like when a trip is added, you also need to add that trip to the driver and passenger's list.  You also missed a major edge-case in testing wave 2.  Wave 1 was well done however.  Think about the `last_trip(driver)` method in `trip_dispatcher.rb`.  That doesn't look like it's in the right place.  Let me know any questions you have.
